### PR TITLE
Scheduler internal NodeTree thread-safe NumNodes

### LINKED ISSUE #70946: [scheduler] NodeTree thread-safety
**Description:**
`NodeTree`, part of the `pkg/scheduler/internal/cache` is exporting a `NumNodes` field which allows to modify its value externally. This also means that this field is not thread-safe.

**Solution:**
Expose a method that returns this value in a thread-safe way and do not export the field.

/kind bug
/sig scheduling

### PR DESCRIPTION
**Pull request by Adirio**
**What type of PR is this?**
/kind cleanup
/sig scheduling

**What this PR does / why we need it**:
Thread-safety for NodeTree

**Which issue(s) this PR fixes**:
Fixes #70946 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
none
```

### CODE REVIEW

**Comment by Adirio** [comment]
Following @k8s-ci-robot suggestion:

/assign @ravisantoshgudimetla

File: pkg/scheduler/internal/cache/node_tree_test.go [hunk 1]
```diff
@@ -116,8 +116,9 @@ func verifyNodeTree(t *testing.T, nt *NodeTree, expectedTree map[string]*nodeArr
 	for _, na := range expectedTree {
 		expectedNumNodes += len(na.nodes)
 	}
-	if nt.NumNodes != expectedNumNodes {
-		t.Errorf("unexpected NodeTree.numNodes. Expected: %v, Got: %v", expectedNumNodes, nt.NumNodes)
+	numNodes := nt.NumNodes()
+	if numNodes != expectedNumNodes {
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/cache/node_tree_test.go]: A neat way:

```suggestion
	if numNodes := nt.NumNodes(); numNodes != expectedNumNodes {
```

File: pkg/scheduler/internal/cache/node_tree.go [hunk 2]
```diff
@@ -184,3 +184,10 @@ func (nt *NodeTree) Next() string {
 		}
 	}
 }
+
+// NumNodes returns the number of nodes.
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/internal/cache/node_tree.go]: In the spirit of what we are doing, I think it's better to call this func as `GetNoOfNodes()`

File: pkg/scheduler/internal/cache/node_tree.go [hunk 3]
```diff
@@ -111,7 +111,7 @@ func (nt *NodeTree) removeNode(n *v1.Node) error {
 					nt.removeZone(zone)
 				}
 				klog.V(5).Infof("Removed node %v in group %v from NodeTree", n.Name, zone)
-				nt.NumNodes--
+				nt.numNodes--
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/internal/cache/node_tree.go]: noit: Shouldn't log statement be after `numNodes--`

File: pkg/scheduler/internal/cache/node_tree_test.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/internal/cache/node_tree_test.go]: Or `expectedNumNodes != nt.NumNodes()`

File: pkg/scheduler/internal/cache/node_tree.go (hunk 2 above)
Comment (Adirio) [reply pkg/scheduler/internal/cache/node_tree.go]: I tried to keep it as similar as it is right now, but if that name fits better I'll change it.
@Huang-Wei what do you think?

File: pkg/scheduler/internal/cache/node_tree.go (hunk 3 above)
Comment (Adirio) [reply pkg/scheduler/internal/cache/node_tree.go]: They are always being done before the changes, I did not modify that

File: pkg/scheduler/internal/cache/node_tree.go (hunk 2 above)
Comment (Huang-Wei) [reply pkg/scheduler/internal/cache/node_tree.go]: Either name is fine to me.

**Comment by Huang-Wei** [comment]
@Adirio the code changes look good to me. Thanks for the PR.

I want to know if you encounter a *real* issue that `NumNodes` are modified in an unsafe way? I know it's unsafe, but want to to evaluate the necessity.

Also cc/ @bsalamat for thoughts.

**Comment by Adirio** [comment]
> I want to know if you encounter a _real_ issue that `NumNodes` are modified in an unsafe way? I know it's unsafe, but want to to evaluate the necessity.

I did not found any, I was inspecting the scheduler code as I'm trying to get a deeper undestanding. Being moved to an internal package, its attack surface its even smaller, but I thought that even if small every attack surface that can be removed should be.

**Review by yastij — APPROVED** [review]
/lgtm

This change SGTM since if the nodeTree and generic_scheduler are executed in different goroutines it might lead to a race condition.

**Review by bsalamat — APPROVED** [review]
Having better thread safety is good, although I don't see a real race issue here.

Thanks, @Adirio!

/lgtm
/approve

**Comment by AishSundar** [comment]
@bsalamat @Huang-Wei @Adirio is this critical for 1.13? we are in code freeze and if possible, I would like to defer this until 1.14. let me know otherwise

/remove-priority critical-urgent

**Comment by Adirio** [comment]
@AishSundar Im not part of the kubernetes team so I do not know if it can be considered a critical bug. It does close an open gate for race conditions. Fetja retested so I can't see the error log for the failed test, but I dont think it will be related to this PR. The PR has both the LGTM and aproved tags so it should be quite easy to get this merged before 1.13-

**Comment by Adirio** [comment]
Tests completed successfully.

**Pull request merged**